README
The jshu archives contain the jshutest.inc shell script "include" file itself
and a sample directory containing a shell script (increment_build.sh) to increment
build numbers in a text file and two different test scripts for it:

* incrbuild_unitTest.sh - This test script is written more along the lines
    of classic developer unit-style tests where tests run individual functions
    inside of increment_build.sh to determine if they are correct.

* incrbuild_funcTest.sh - This test script is written more like a functional
    or integration test - where the internal components are presumed to already
    have been tested elsewhere and we now wish to test the operation of the script
    as a whole. Thus, this script provides increment_build.sh with data, runs it,
    and evaluates the resultant data after execution is done.


CHANGELOG
Version 1.0.0-20
* Add tests duration in tests reports
* Use the results/xunit path for XML results
* Refactor XML with timestamps, failures, hostname
* Return error level in finish steps
* Use 'Tests.*' naming for XML result files
* Always adapt the package name to a file name
* Improve tests outputs with colors and spacing
* Trim extended and colored chars from execution outputs
* Extract jshu_errmsg and test_title from the output
* Handle errors inside test functions and verbose output
* Implement optional "test_title" variable in test
* Support passing a formatted tests suite name
* Apply shfmt formatting Bash codestyle to all files
* Enforce faulty result file before proper jshuFinalize call
* Support wrapper.sh --test handlings with enforced failures

Version 1.0.0-19
* Removed testcase property (assertions) that Jenkins no longer supports.
* Limited elapsed time to 3 decimals because the Jenkins time validation is now broken.

Version 1.0.0-18
* Fixed up function ordering errors and a dangling quote.

Version 1.0.0-17
* Patches submitted by ireicht to fix issues with file paths that have embedded spaces and to fix a segfault in the sample/incrbuild_unitTest.sh
* Modification to the patched sample/incrbuild_unitTest.sh to change the name ofa "dynamic" test function so that it will not also be a "discovered" test function (so the test will not run twice).

Version 1.0.0-16
* Check to make sure that any test function names that were provided (as opposed
  to discovered) actually exist in the test script.
  
Version 1.0.0-15
* Add the ability to specify test function names in a variable (jshuTestFunctions)
  that will be run before the discovered functions (those whose names end with Test).
  See the test/addhocFunctionNamesTest.sh script or the sample/incrbuild_unitTest.sh
  script for an example of use.
  
Version 1.0.0-14
* Add two possible return values for a test - ${jshuFAILABORT} and
  ${jshuERRORABORT}. These allow you to mark failures and errors in a test
  that are so severe that the rest of the tests in the test suite should not
  be run.
Version 1.0.0-13
* Modify calling of jshuSetup and jshuTeardown functions so that if they fail 
  for some reason, the failure will be caught and reported.
  
Version 1.0.0-12
* Prevent test result writing from erroring out when you call a test script
  from a test script - as long as you are in a bash shell - because only the bash
  shell provides a means of getting a subshell pid.

Version 1.0.0-11
* Bring back build 9 and then protect against missing capture file.

Version 1.0.0-10
* bad release - reverts back to 1.0.0-5

Version 1.0.0-9
* Filter binary (unprintable) characters out of the CDATA stdout capture to
  prevent malformed XML.
  
Version 1.0.0-8
* Fix problem of malformed xml when a jshu_errmsg string has an embedded quote mark. 
  Convert it to html entity &quot; .

Version 1.0.0-7
* Add support for skipped test messages now that I know Jenkins can display them.
  Simply assign your skip message to jshu_errmsg before the ${jshuTEST_SKIP}.

Version 1.0.0-6
* Change directory back to where the test script was started from before executing
  each test function.
* Miscellaneous code cleanup and commenting.


tests/fail_FailAbortTest.sh
#!/bin/bash 

# source the unit test for scripts functions
. ../jshutest.inc

# see to it that the setup function fails
jshuSetup() {
    return 0
}
##############################################################
# unit test functions
AbortingTest() {
    return ${jshuFAILABORT}
}

AlwaysPassTest() {
	return ${jshuPASS}
}




##############################################################
# main
##############################################################
# initialize testsuite
jshuInit

# run unit tests on script
jshuRunTests

# result summary
jshuFinalize

echo Done.
echo
let tot=failed+errors
exit $tot


tests/fail_jshuSetupTest.sh
#!/bin/bash 

# source the unit test for scripts functions
. ../jshutest.inc

# see to it that the setup function fails
jshuSetup() {
    return 1
}
##############################################################
# unit test functions
AlwaysPassTest() {
	return ${jshuPASS}
}




##############################################################
# main
##############################################################
# initialize testsuite
jshuInit

# run unit tests on script
jshuRunTests

# result summary
jshuFinalize

echo Done.
echo
let tot=failed+errors
exit $tot


tests/fail_jshuTeardownTest.sh
#!/bin/bash 

# source the unit test for scripts functions
. ../jshutest.inc

jshuSetup() {
    return 0
}
# see to it that the teardown function fails
jshuTeardown() {
    return 1
}

##############################################################
# unit test functions
AlwaysPassTest() {
	return ${jshuPASS}
}




##############################################################
# main
##############################################################
# initialize testsuite
jshuInit

# run unit tests on script
jshuRunTests

# result summary
jshuFinalize

echo Done.
echo
let tot=failed+errors
exit $tot


wrapper.sh
#!/bin/bash

# Variables
jshu=$(dirname $(readlink -f "${BASH_SOURCE:-${0}}"))
result=1

# Wrapper modes
case "${1}" in

# Load mode
--load)
  # Configure bash
  set +e
  set +x

  # Import jshu
  source "${jshu}/jshutest.inc"

  # Configure tests package
  jshu_pkgname="${2}"

  # Start tests
  if [ ! -z "${WORKSPACE}" ] && [ -z "${BUILDDIR}" ]; then
    BUILDDIR=${WORKSPACE} jshuInit "${3}"
  else
    jshuInit "${3}"
  fi

  # Declare Test function
  function Test() {
    # Variables
    local result

    # Create test runner
    source <(
      cat <<END_OF_TEST
function TestRunner()
{
  # Test title
  test_title="${1}";

  # Test executions
  $(cat)
}
END_OF_TEST
    )

    # Run test
    jshu_run_test TestRunner
    result=${?}

    # Cleanup
    unset TestRunner

    # Result
    return "${result}"
  }

  # Result
  result=0
  ;;

# Test mode
--test)
  # Load jshu wrapper
  source "${jshu}/wrapper.sh" --load "${2}" "${3}"

  # Add test name to the file name
  jshu_test_tag=${4//./_}
  jshu_test_tag=${jshu_test_tag// /_}
  jshu_suite=${jshu_suite}.${jshu_test_tag}

  # Run single test
  Test "${4}" <<§
  $(cat)
§
  result=${?}

  # Finish jshu wrapper
  source "${jshu}/wrapper.sh" --finish
  ;;

# Finish mode
--finish)
  # Finalize tests
  jshuFinalize
  result=${?}
  ;;

esac

# Result
(exit "${result}")


sample/incrbuild_funcTest.sh
#!/bin/bash

# source the unit test for scripts functions
. ../jshutest.inc

# helper function
# remove host and date lines from buildnumber.txt file
stripHostDate() {
  source="$1"
  dest="$2"
  grep -v '^BuildHost:' $source | grep -v '^Date' >$dest
}

# In this one, we don't source the other script because
# we're not going to test pieces of it. We're just going
# to run the script and evaluate what it does.
##############################################################
# unit test functions
emptyBldNumFile_Test() {
  tmpfiles_used="./ldempty_test.txt ./ldempty_test1.txt"
  /bin/rm -f $tmpfiles_used

  ./increment_build.sh build ./ldempty_test.txt

  stripHostDate ./ldempty_test.txt ./ldempty_test1.txt
  if ! diff ./ldempty_test1.txt data/empty.txt; then
    /bin/rm -f $tmpfiles_used
    return ${jshuFAIL}
  fi

  # cleanup files
  /bin/rm -f $tmpfiles_used
  return ${jshuPASS}
}

myprodFile_superTest() {
  tmpfiles_used="./myprod_test.txt ./myprod_test1.txt"
  /bin/rm -f $tmpfiles_used

  cp data/myprod.txt ./myprod_test.txt
  ./increment_build.sh super ./myprod_test.txt

  stripHostDate ./myprod_test.txt ./myprod_test1.txt
  if ! diff ./myprod_test1.txt data/myprod.sup.txt; then
    /bin/rm -f $tmpfiles_used
    return ${jshuFAIL}
  fi

  # cleanup files
  /bin/rm -f $tmpfiles_used
  return ${jshuPASS}
}

myprodFile_minorTest() {
  tmpfiles_used="./myprod_test.txt ./myprod_test1.txt"
  /bin/rm -f $tmpfiles_used

  cp data/myprod.txt ./myprod_test.txt
  ./increment_build.sh minor ./myprod_test.txt

  stripHostDate ./myprod_test.txt ./myprod_test1.txt
  if ! diff ./myprod_test1.txt data/myprod.min.txt; then
    /bin/rm -f $tmpfiles_used
    return ${jshuFAIL}
  fi

  # cleanup files
  /bin/rm -f $tmpfiles_used
  return ${jshuPASS}
}

##############################################################
# main
##############################################################
# initialize testsuite
jshuInit

# run unit tests on script
jshuRunTests

# result summary
jshuFinalize

echo Done.
echo
let tot=failed+errors
exit $tot


sample/incrbuild_unitTest.sh
#!/bin/bash

# source the unit test for scripts functions
. ../jshutest.inc

# source our script that we want to test features of
. ./increment_build.sh

jshuSetup() {
  # set this variable to a space-delimited list of non-standard test function
  # names that you want to run as tests. They will run before the *Test
  # functions. It will be used by the jshuRunTests function, so if you
  # want to use it, you should assign the value to it before calling
  # jshuRunTests in the bottom (boilerplate) section.
  jshuTestFunctions="parseVersionFunc getBldFileName_noParam"
}

##############################################################
# unit test functions
parseVersionFunc() {
  VERSION="1.5.7"
  parseVersion
  if [ "$VERSION_SUPER" != "1" ]; then
    return ${jshuFAIL}
  fi
  if [ "$VERSION_MAJOR" != "5" ]; then
    return ${jshuFAIL}
  fi
  if [ "$VERSION_MINOR" != "7" ]; then
    return ${jshuFAIL}
  fi
  return ${jshuPASS}
}

parseVersionShortTest() {
  VERSION="1.5"
  parseVersion
  echo $VERSION_SUPER.$VERSION_MAJOR.$VERSION_MINOR
  if [ "$VERSION_SUPER" != "1" ]; then
    jshu_errmsg="\$VERSION_SUPER ($VERSION_SUPER) != 1"
    return ${jshuFAIL}
  fi
  if [ "$VERSION_MAJOR" != "5" ]; then
    jshu_errmsg="\$VERSION_MAJOR ($VERSION_MAJOR) != 5"
    return ${jshuFAIL}
  fi
  if [ "$VERSION_MINOR" != "" ]; then
    jshu_errmsg="\$VERSION_MINOR ($VERSION_MINOR) != \"\""
    return ${jshuFAIL}
  fi
  return ${jshuPASS}
}

incrSuperTest() {
  VERSION="1.5.7"
  BUILDNUM=3
  EPOCH=7003
  parseVersion
  incrSuper
  if [ ${VERSION_SUPER} -ne 2 ]; then
    jshu_errmsg="\$VERSION_SUPER ($VERSION_SUPER) != 1"
    return ${jshuFAIL}
  fi
  if [ "$VERSION_MAJOR" != "0" ]; then
    jshu_errmsg="\$VERSION_MAJOR ($VERSION_MAJOR) != 0"
    return ${jshuFAIL}
  fi
  if [ "$VERSION_MINOR" != "0" ]; then
    jshu_errmsg="\$VERSION_MINOR ($VERSION_MINOR) != 0"
    return ${jshuFAIL}
  fi
  if [ "$BUILDNUM" != "1" ]; then
    jshu_errmsg="\$BUILDNUM ($BUILDNUM) != 1"
    return ${jshuFAIL}
  fi
  if [ "$EPOCH" != "7004" ]; then
    jshu_errmsg="\$EPOCH ($EPOCH) != 7004"
    return ${jshuFAIL}
  fi
  return ${jshuPASS}
}

incrMajorTest() {
  VERSION="1.5.7"
  BUILDNUM=3
  EPOCH=7003
  parseVersion
  incrMajor
  if [ ${VERSION_SUPER} -ne 1 ]; then
    jshu_errmsg="\$VERSION_SUPER ($VERSION_SUPER) != 1"
    return ${jshuFAIL}
  fi
  if [ "$VERSION_MAJOR" != "6" ]; then
    jshu_errmsg="\$VERSION_MAJOR ($VERSION_MAJOR) != 6"
    return ${jshuFAIL}
  fi
  if [ "$VERSION_MINOR" != "0" ]; then
    jshu_errmsg="\$VERSION_MINOR ($VERSION_MINOR) != 0"
    return ${jshuFAIL}
  fi
  if [ "$BUILDNUM" != "1" ]; then
    jshu_errmsg="\$BUILDNUM ($BUILDNUM) != 1"
    return ${jshuFAIL}
  fi
  if [ "$EPOCH" != "7004" ]; then
    jshu_errmsg="\$EPOCH ($EPOCH) != 7004"
    return ${jshuFAIL}
  fi
  return ${jshuPASS}
}

incrMinorTest() {
  VERSION="1.5.7"
  BUILDNUM=3
  EPOCH=7003
  parseVersion
  incrMinor
  if [ ${VERSION_SUPER} -ne 1 ]; then
    jshu_errmsg="\$VERSION_SUPER ($VERSION_SUPER) != 1"
    return ${jshuFAIL}
  fi
  if [ "$VERSION_MAJOR" != "5" ]; then
    jshu_errmsg="\$VERSION_MAJOR ($VERSION_MAJOR) != 5"
    return ${jshuFAIL}
  fi
  if [ "$VERSION_MINOR" != "8" ]; then
    jshu_errmsg="\$VERSION_MINOR ($VERSION_MINOR) != 8"
    return ${jshuFAIL}
  fi
  if [ "$BUILDNUM" != "1" ]; then
    jshu_errmsg="\$BUILDNUM ($BUILDNUM) != 1"
    return ${jshuFAIL}
  fi
  if [ "$EPOCH" != "7004" ]; then
    jshu_errmsg="\$EPOCH ($EPOCH) != 7004"
    return ${jshuFAIL}
  fi
  return ${jshuPASS}
}

incrBuildTest() {
  VERSION="1.5.7"
  BUILDNUM=3
  EPOCH=7003
  parseVersion
  incrBuild
  if [ ${VERSION_SUPER} -ne 1 ]; then
    jshu_errmsg="\$VERSION_SUPER ($VERSION_SUPER) != 1"
    return ${jshuFAIL}
  fi
  if [ "$VERSION_MAJOR" != "5" ]; then
    jshu_errmsg="\$VERSION_MAJOR ($VERSION_MAJOR) != 5"
    return ${jshuFAIL}
  fi
  if [ "$VERSION_MINOR" != "7" ]; then
    jshu_errmsg="\$VERSION_MINOR ($VERSION_MINOR) != 7"
    return ${jshuFAIL}
  fi
  if [ "$BUILDNUM" != "4" ]; then
    jshu_errmsg="\$BUILDNUM ($BUILDNUM) != 4"
    return ${jshuFAIL}
  fi
  if [ "$EPOCH" != "7004" ]; then
    jshu_errmsg="\$EPOCH ($EPOCH) != 7004"
    return ${jshuFAIL}
  fi
  return ${jshuPASS}
}

getBldFileName_noParam() {
  bfname=$(getBldNumFilename)
  if [ "$bfname" != "buildnumber.txt" ]; then
    return ${jshuFAIL}
  fi
  return ${jshuPASS}
}

getBldFileName_paramTest() {
  bfname=$(getBldNumFilename mybldnum.txt)
  if [ "$bfname" != "mybldnum.txt" ]; then
    return ${jshuFAIL}
  fi
  return ${jshuPASS}
}

createBldNumFile_newTest() {
  tmpfiles_used="./crbldfile_test.txt"
  /bin/rm -f $tmpfiles_used

  createBldNumFile ./crbldfile_test.txt
  if [ ! -e ./crbldfile_test.txt ]; then
    jshu_errmsg="[ ! -e ./crbldfile_test.txt ] failed"
    return ${jshuFAIL}
  fi
  /bin/rm -f $tmpfiles_used
  return ${jshuPASS}
}

loadBldNumFile_emptyTest() {
  tmpfiles_used="./ldbldfile_test.txt"
  /bin/rm -f $tmpfiles_used

  createBldNumFile ./ldbldfile_test.txt
  parseVersion
  [ ! -e ./ldbldfile_test.txt ] && return ${jshuFAIL}
  loadBldNumFile ./ldbldfile_test.txt
  /bin/rm -f $tmpfiles_used

  if [ "${PRODNAME}" = "Unknown product" ]; then
    jshu_errmsg="\$PRODNAME ($PRODNAME) != Unknown product"
    return ${jshuFAIL}
  fi
  if [ "$PKGNAME" != "unknown_pkg" ]; then
    jshu_errmsg="\$PKGNAME ($PKGNAME) != unknown_pkg"
    return ${jshuFAIL}
  fi
  if [ "$VERSION" != "0.0.0" ]; then
    jshu_errmsg="\$VERSION ($VERSION) != 0.0.0"
    return ${jshuFAIL}
  fi
  if [ "$BUILDNUM" != "0" ]; then
    jshu_errmsg="\$BUILDNUM ($BUILDNUM) != 0"
    return ${jshuFAIL}
  fi
  if [ "$EPOCH" != "0" ]; then
    jshu_errmsg="\$EPOCH ($EPOCH) != 0"
    return ${jshuFAIL}
  fi
  return ${jshuPASS}
}

loadBldNumFileTest() {
  tmpfiles_used="./ldbldfile1_test.txt"
  /bin/rm -f $tmpfiles_used

  PRODNAME="My sample product"
  PKGNAME="sample_prod"
  VERSION="1.2.3"
  parseVersion
  BUILDNUM="145"
  EPOCH="7009125"
  BUILDHOST="ananke"
  BUILDDATE="Sun Jan 26 15:57:13 CST 2014"
  writeBldNumFile ./ldbldfile1_test.txt
  unset PRODNAME PKGNAME VERSION BUILDNUM EPOCH

  loadBldNumFile ldbldfile1_test.txt
  /bin/rm -f $tmpfiles_used
  if [ "${PRODNAME}" != "My sample product" ]; then
    jshu_errmsg="\$PRODNAME ($PRODNAME) != My sample product"
    return ${jshuFAIL}
  fi
  if [ "${PKGNAME}" != "sample_prod" ]; then
    jshu_errmsg="\$PKGNAME (${PKGNAME}) != sample_prod"
    return ${jshuFAIL}
  fi
  if [ "${VERSION}" != "1.2.3" ]; then
    jshu_errmsg="\$VERSION ($VERSION) != 1.2.3"
    return ${jshuFAIL}
  fi
  if [ "${BUILDNUM}" != "145" ]; then
    jshu_errmsg="\$BUILDNUM ($BUILDNUM) != 145"
    return ${jshuFAIL}
  fi
  if [ "${EPOCH}" != "7009125" ]; then
    jshu_errmsg="\$EPOCH ($EPOCH) != 7009125"
    return ${jshuFAIL}
  fi
  return ${jshuPASS}
}

badErrorMsgTest() {
  jshu_errmsg="This is a \" quote mark"
  ${jshuTEST_SKIP}
}

unfinishedTest() {
  jshu_errmsg="This test has not been written yet"
  ${jshuTEST_SKIP}
}

##############################################################
# main
##############################################################
# initialize testsuite
jshuInit

# run unit tests on script
jshuRunTests

# result summary
jshuFinalize

echo Done.
echo
let tot=failed+errors
exit $tot


sample/increment_build.sh
#!/bin/bash

# increment_build.sh
#
# This script allows for simple maintenance of an informational text
# file containing information useful for specification in an rpm spec file.
#
# For example, one such file might contain:
#	ProductName: Alidnis Network Defense System
#	PkgName: alidnis
#	Version: 1.0.0
#	BuildNumber: 1
#	Epoch: 64
#	BuildHost: ananke.shadowguard
#	Date: Wed Sep  5 13:26:20 CDT 2012
#
# The BUILDNUM_TXT variable (below) has the name of your buildnumber.txt
# file (including any necessary directory info). If the
# BUILDNUM_TXT file does not exist, it will be created
# with dummy values. You can then edit the BUILDNUM_TXT
# file to provide product and package names, etc.
#
# Usage:
#	increment_build.sh incr_option [alternate buildnumber.txt file]
#
# where incr_option is one of the following:
#   super - the first number of the version triple a.b.c
#   major - the second number of the version triple a.b.c
#   minor - the third number of the version triple a.b.c
#   build - the build number
# and the alternate buildnumber.txt file is the name of the file that
# you wish to keep your info in different from the default "buildnumber.txt".
#
# When the incr_option is:
#   super - the super number is incremented,
#           the major and minor numbers are reset to zero,
#           and build number is reset to one
#   major - the major number is incremented,
#           the minor number is reset to zero,
#           and build number is reset to one
#   minor - the minor number is incremented,
#           and build number is reset to one
#   build - the build number is incremented
#
# The epoch number is always incremented and never reset.
#
# One way you can use this is in the Makefile where you
# build your rpm. Define the following to get access to
# the values (add directory info to "buildnumber.txt"
# if necessary):
#   PRODNAME = $(shell /bin/grep ^ProductName: buildnumber.txt | /bin/cut -d: -f2)
#   PKGNAME = $(shell /bin/grep ^PkgName: buildnumber.txt | /bin/awk '{print $$2}')
#   VERSION = $(shell /bin/grep ^Version: buildnumber.txt | /bin/awk '{print $$2}')
#   RELEASE = $(shell /bin/grep ^BuildNumber: buildnumber.txt | /bin/awk '{print $$2}')
#   EPOCH = $(shell /bin/grep ^Epoch: buildnumber.txt | /bin/awk '{print $$2}')
#
# Add the parameters to your rpmbuild command to pass them in. For example:
#	cd $(RPMBUILD)/SPECS; \
#	rpmbuild --define '_topdir $(RPMBUILD)' \
#		--define 'prodname $(PRODNAME)' \
#		--define 'pkgname $(PKGNAME)' \
#		--define 'version $(VERSION)' \
#		--define 'release $(RELEASE)' \
#		--define 'epoch $(EPOCH)' \
#		-bb $(PKGNAME).spec
#
# And finally, use the variables you defined in the rpmbuild
# command inside your spec file. For example:
#    Summary: %{prodname} installation package
#    Name: %{pkgname}
#    Version: %{version}
#    Release: %{release}
#    BuildArch: i686
#    Epoch: %{epoch}
#
#
#
#-----------------------------------------------------------
# License: Simplified BSD:
#
# Copyright 2012, Dolores Scott. All rights reserved.
#
# Redistribution and use in source and binary forms, with or without modification, are
# permitted provided that the following conditions are met:
#
#   1. Redistributions of source code must retain the above copyright notice, this list of
#      conditions and the following disclaimer.
#
#   2. Redistributions in binary form must reproduce the above copyright notice, this list
#      of conditions and the following disclaimer in the documentation and/or other materials
#      provided with the distribution.

# THIS SOFTWARE IS PROVIDED BY Dolores Scott ``AS IS'' AND ANY EXPRESS OR IMPLIED
# WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
# FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL Dolores Scott OR
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
# ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
# NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

BUILDNUM_TXT=./buildnumber.txt

# global variables used in the script:
#    PRODNAME, PKGNAME, VERSION, BUILDNUM, EPOCH,
#    BUILDHOST, BUILDDATE

# get the name of the buildnumber file to use
getBldNumFilename() {
  if [ $# -lt 1 ]; then
    filename="buildnumber.txt"
  else
    filename="$1"
  fi
  if [ ! -n "$filename" ]; then
    filename="buildnumber.txt"
  fi
  echo $filename
  unset filename
}

writeBldNumFile() {
  filename="$1"

  echo "ProductName:$PRODNAME" >"${filename}"
  echo "PkgName: $PKGNAME" >>"${filename}"
  if [ -n "$VERSION_MINOR" ]; then
    ver="$VERSION_SUPER.$VERSION_MAJOR.$VERSION_MINOR"
  else
    ver="$VERSION_SUPER.$VERSION_MAJOR"
  fi
  echo "Version: $ver" >>"${filename}"
  echo "BuildNumber: $BUILDNUM" >>"${filename}"
  echo "Epoch: $EPOCH" >>"${filename}"
  echo "BuildHost: $BUILDHOST" >>"${filename}"
  echo "Date: $BUILDDATE" >>"${filename}"
  unset filename ver
}

parseVersion() {
  VERSION_SUPER=${VERSION%%.*}
  VERSION_MINOR=${VERSION##*.}
  vermajmin=${VERSION#*.}
  VERSION_MAJOR=${vermajmin%.*}
  if [ "$vermajmin" = "$VERSION_MAJOR" ]; then
    # we don't actually HAVE a minor version
    VERSION_MINOR=""
  fi
}

# only creates a file if it does not already exist!!
# will not overwrite an existing file
createBldNumFile() {
  filename="$1"
  if [ ! -n "$filename" ]; then
    return 1
  fi
  if [ ! -e "$filename" ]; then
    /bin/echo "Could not find $filename"
    BLDIR=${filename%/*}
    if [ "$BLDIR" != "$filename" ]; then
      if [ ! -d $BLDIR ]; then
        echo "$BLDIR directory does not exist... aborting" >&2
        return 1
      fi
    fi
    /bin/echo "   Creating it."
    PRODNAME=" Unknown product" # yes, the initial space is deliberate
    PKGNAME="unknown_pkg"
    VERSION="0.0.0"
    parseVersion
    BUILDNUM="0"
    EPOCH="0"
    BUILDHOST=$(/bin/uname -a | /bin/awk '{ print $2}')
    BUILDDATE=$(/bin/date)
    writeBldNumFile "$filename"
  fi
  unset filename
}

loadBldNumFile() {
  filename="$1"
  if [ ! -e "$filename" ]; then
    return 1
  fi

  PRODNAME=$(/bin/grep ^ProductName "$filename" | /bin/cut -d: -f2)
  PKGNAME=$(/bin/grep ^PkgName "$filename" | /bin/awk '{print $2}')
  VERSION=$(/bin/grep ^Version "$filename" | /bin/awk '{print $2}')
  BUILDNUM=$(/bin/grep ^BuildNumber "$filename" | /bin/awk '{print $2}')
  EPOCH=$(/bin/grep "^Epoch" "$filename" | /bin/awk '{print $2}')
  unset filename
}

incrSuper() {
  echo -e "\nIncrementing version super...\n"
  let VERSION_SUPER=$VERSION_SUPER+1
  VERSION_MAJOR=0
  VERSION_MINOR=0
  let BUILDNUM=1
  let EPOCH=$EPOCH+1
}

incrMajor() {
  echo -e "\nIncrementing version major...\n"
  let VERSION_MAJOR=$VERSION_MAJOR+1
  VERSION_MINOR=0
  let BUILDNUM=1
  let EPOCH=$EPOCH+1
}

incrMinor() {
  set -x -a -v
  echo -e "\nIncrementing version minor...\n"
  if [ -z "VERSION_MINOR" ]; then
    VERSION_MINOR=0
  else
    let VERSION_MINOR=$VERSION_MINOR+1
  fi
  let BUILDNUM=1
  let EPOCH=$EPOCH+1
  set +x +a +v
}

incrBuild() {
  echo -e "\nIncrementing build number...\n"
  let BUILDNUM=$BUILDNUM+1
  let EPOCH=$EPOCH+1
}

usage() {
  echo -e "USAGE:\n\t$0 super|major|minor|build [alternate buildnumber.txt file]"
}
#############################################################################################
# main
#    (Only execute the following if we were called by our real name - i.e. as a
#     main script. Otherwise assume we've been sourced for testing of the functions.)
#
if [ ${0##*/} == "increment_build.sh" ]; then
  echo $@

  if [ -z "$1" ] || [ "$1" != "major" ] \
    && [ "$1" != "minor" ] && [ "$1" != "build" ] \
    && [ "$1" != "super" ]; then
    /bin/echo -e "\nERROR!  Invalid or missing parameter."
    usage
    exit 1
  fi
  opt="$1"
  if [ $# -lt 2 ]; then
    BUILDNUM_TXT=$(getBldNumFilename)
  else
    BUILDNUM_TXT=$(getBldNumFilename $2)
  fi
  createBldNumFile $BUILDNUM_TXT
  loadBldNumFile $BUILDNUM_TXT

  VERSION_SUPER=${VERSION%%.*}
  VERSION_MINOR=${VERSION##*.}
  vertmp=${VERSION#*.}
  VERSION_MAJOR=${vertmp%.*}
  case $opt in
    "super")
      incrSuper
      ;;
    "major")
      incrMajor
      ;;
    "minor")
      incrMinor
      ;;
    "build")
      incrBuild
      ;;
    *)
      echo "ERROR: Unknown option \"$opt\""
      usage
      exit 1
      ;;
  esac

  BUILDHOST=$(/bin/uname -a | /bin/awk '{ print $2}')
  BUILDDATE=$(/bin/date)

  writeBldNumFile ${BUILDNUM_TXT}
fi
